### PR TITLE
Less asserts, more catching and raising errors

### LINKED ISSUE #36: Reconsider asserts
I really like the Defensive Programming unit, but I'm of the opinion that you may want to consider teaching `assert` as a tool for validating user input. For one, explicit errors like `ValueError` are more descriptive of the problem rather than the generic `AssertionError`. The other problem is that if you run python as `python -O` or `python -OO` (which is short for setting `PYTHONOPTIMIZE=1`), all `assert` statements are ignored/compiled out. In general, assertions are supposed to be used to double check conditions that should never occur, not for validation of user input. Some other posts around this:
* https://dev.to/yloganathan/do-not-abuse-the-assert-1cfm
* http://archive.is/5GfiG

Just my $0.02.

### PR DESCRIPTION
**Pull request by DamienIrving**
Resolves #36

### CODE REVIEW

**Comment by DamienIrving** [comment]
@hot007 @sarahymurphy - As maintainers on these PyAOS lessons I just wanted to let you know that in response to feedback from the workshops we did for the AMS and ICSHMO (see #36) I've done a re-write of the defensive programming lesson. 

Feel free to post comments/feedback on this PR.

If you'd like to see what the new lesson would actually look like on the web, you can view my fork at https://damienirving.github.io/python-aos-lesson/08-defensive/index.html

File: _episodes/08-defensive.md [hunk 1]
```diff
@@ -237,40 +342,250 @@ $ git push origin master
...
+    
+logging.basicConfig(level=logging.debug, filename='log.txt')) 
+
+rh_data = np.array([1.5, 20.4, 100.1, 76.3, 54.4])
+rh_max = rh_data.max()
+logging.debug(f'The maximum relative humidity was {rh_max}%')
+~~~
+{: .language-python}
+
+(The for loop is needed to turn off the background logging the notebook does itself.
```
Comment (hot007) [inline _episodes/08-defensive.md:421]: Suggested addition:
By setting the logging level to "debug", our output "log.txt" file will now capture all logging information with a flag of 'debug' or higher - that is, all logging outputs will be written to our log file.

**Review by hot007 — COMMENTED** [review]
This looks really good @DamienIrving , a couple of minor corrections/suggestions but I'm happy to approve and you can implement the changes or not. 

I like the new approach of the lesson, I think this is an improvement. While it adds cognitive load by showing a couple of ways to handle errors, I think it does it in a way that participants will understand, and can be supported in a teaching environment by talking through what choices you might make about how to handle error checking in a particular situation, and that at the end of the day any sort of automated error handling is vastly better than none!!

**Comment by hot007** [comment]
Sorry for the slow feedback by the way, was very busy with another deadline!

**Comment by DamienIrving** [comment]
Thanks, @hot007! I've implemented your suggestions in https://github.com/carpentries-lab/python-aos-lesson/commit/3a83ac687c0081275670427a74f78bc7772e9938

**Pull request merged**